In [53]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Imputación de nulos usando métodos avanzados estadísticos
# -----------------------------------------------------------------------
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# Gestión de los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings('ignore')

In [54]:
# Extracción de los datos iniciales
df = pd.read_csv('files/hr.csv')
df.head(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0
2,37.0,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,lABORATORY tECHNICIAN,3.0,Single,2090.0,2396,6,Y,Yes,15,3,2,NaN,0,7,3.0,3,0,0,0,0.0
3,33.0,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,rESEARCH sCIENTIST,3.0,Married,2909.0,23159,1,Y,Yes,11,3,3,80.0,0,8,3.0,3,8,7,3,0.0
4,27.0,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,lABORATORY tECHNICIAN,2.0,Married,3468.0,16632,9,Y,No,12,3,4,80.0,1,6,3.0,3,2,2,2,2.0


In [55]:
# Información de columnas y tipos de variables del dataset
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1401 non-null   float64
 1   Attrition                 1474 non-null   str    
 2   BusinessTravel            1357 non-null   str    
 3   DailyRate                 1474 non-null   int64  
 4   Department                1445 non-null   str    
 5   DistanceFromHome          1474 non-null   int64  
 6   Education                 1474 non-null   int64  
 7   EducationField            1416 non-null   str    
 8   EmployeeCount             1474 non-null   int64  
 9   EmployeeNumber            1474 non-null   int64  
 10  EnvironmentSatisfaction   1474 non-null   int64  
 11  Gender                    1474 non-null   str    
 12  HourlyRate                1474 non-null   int64  
 13  JobInvolvement            1474 non-null   int64  
 14  JobLevel           

In [56]:
# Comprobación de columnas y filas
df.shape

(1474, 35)

In [57]:
# Búsqueda de columnas con valores nulos
nulos = df.isnull().sum()
nulos = nulos[nulos>0]
nulos

Age                       73
BusinessTravel           117
Department                29
EducationField            58
JobSatisfaction           29
MaritalStatus            132
MonthlyIncome             14
OverTime                  44
StandardHours            164
TrainingTimesLastYear     88
YearsWithCurrManager     148
dtype: int64

In [58]:
# Comprobación de porcentaje de nulos
nulos_pct = round(df.isnull().sum()/df.shape[0]*100, 2)
nulos_pct = nulos_pct[nulos_pct>0]
nulos_pct

Age                       4.95
BusinessTravel            7.94
Department                1.97
EducationField            3.93
JobSatisfaction           1.97
MaritalStatus             8.96
MonthlyIncome             0.95
OverTime                  2.99
StandardHours            11.13
TrainingTimesLastYear     5.97
YearsWithCurrManager     10.04
dtype: float64

In [59]:
# Comprobación de duplicados y verificacióin de que los numeros de empleados están repetidos
print(df.duplicated().sum())
print(df['EmployeeNumber'].duplicated().sum())

4
4


In [60]:
# Visualización de las filas con EmployeeNumber duplicado
# y se ordenan por EmployeeNumber para ver juntas las repeticiones
df[df['EmployeeNumber'].duplicated(keep=False)].sort_values('EmployeeNumber')

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
67,45.0,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,rESEARCH sCIENTIST,1.0,Divorced,9724.0,18787,2,Y,No,17,3,3,NaN,1,25,2.0,3,1,0,0,0.0
1473,45.0,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,rESEARCH sCIENTIST,1.0,Divorced,9724.0,18787,2,Y,No,17,3,3,NaN,1,25,2.0,3,1,0,0,0.0
1471,53.0,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,4,Female,57,4,2,mANUFACTURING dIRECTOR,1.0,Divorced,4450.0,26250,1,Y,No,11,3,3,NaN,2,5,3.0,3,4,2,1,3.0
184,53.0,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,4,Female,57,4,2,mANUFACTURING dIRECTOR,1.0,Divorced,4450.0,26250,1,Y,No,11,3,3,NaN,2,5,3.0,3,4,2,1,3.0
1470,28.0,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,4,Male,84,3,2,sALES eXECUTIVE,1.0,Single,8463.0,23490,0,Y,No,18,3,4,NaN,0,6,4.0,3,5,4,1,NaN
1041,28.0,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,4,Male,84,3,2,sALES eXECUTIVE,1.0,Single,8463.0,23490,0,Y,No,18,3,4,NaN,0,6,4.0,3,5,4,1,NaN
1222,24.0,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,hUMAN rESOURCES,3.0,Married,1555.0,11585,1,Y,No,11,3,3,80.0,1,1,2.0,3,1,0,0,0.0
1472,24.0,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,hUMAN rESOURCES,3.0,Married,1555.0,11585,1,Y,No,11,3,3,80.0,1,1,2.0,3,1,0,0,0.0


In [61]:
# Verificación de si existen valores negativos en el dataset
(df.select_dtypes(include='number') < 0).any().any()

np.False_

**Comprobaciones realizadas:**

* Los títulos de columnas siguen la misma estética.
* `JobRol` tiene los textos con las mayúsculas invertidas.
* `BusinessTravel` tiene guión bajo como separación en vez de espacios.
* Hay 4 filas duplicadas. Se confirma con el número de empleado.
* No hay números negativos en ninguna columna.

**Tareas pendientes:**

- [_] Cambiar `JobRol`y `BusinessTravel` para continuidad de la estética
- [_] Eliminación de duplicados.

In [62]:
# Se comprueba que los valores de Attrition sean solo "Yes" o "No". No tiene nulos
df['Attrition'].unique()

<StringArray>
['Yes', 'No']
Length: 2, dtype: str

In [63]:
# Education es un valor numérico que mide la escala de estudios, se comprueba los valores comprendidos. No tiene nulos
df['Education'].unique()

array([2, 1, 4, 3, 5])

In [64]:
# Solo debe devolver valor "1". No aporta valor
df['EmployeeCount'].unique()

array([1])

In [65]:
# Encuesta de satisfacción. Escala de valores entre 1 y 4. No tiene nulos
df['EnvironmentSatisfaction'].unique()

array([2, 3, 4, 1])

In [66]:
# Se comprueba que haya solo dos valores. No tiene nulos
df['Gender'].unique()

<StringArray>
['Female', 'Male']
Length: 2, dtype: str

In [67]:
# Encuesta de satisfacción. Escala de valores entre 1 y 4. No tiene nulos
df['JobInvolvement'].unique()

array([3, 2, 4, 1])

In [68]:
# Jerarquía de la empresa. Valores entre 1 y 5. No tiene nulos
df['JobLevel'].unique()

array([2, 1, 3, 4, 5])

In [69]:
# Encuesta de satisfacción. Escala de valores entre 1 y 4. Tiene nulos
df['JobSatisfaction'].unique()

array([ 4.,  2.,  3.,  1., nan])

In [70]:
# Comprobación de que solo existan cuatro valores y nulos
df['MaritalStatus'].unique()

<StringArray>
['Single', 'Married', 'Divorced', nan, 'Marreid']
Length: 5, dtype: str

In [71]:
# Comprobación del número de valores
df['NumCompaniesWorked'].unique()

array([8, 1, 6, 9, 0, 4, 5, 2, 7, 3])

In [72]:
# Todos los valores en "Y"
df['Over18'].unique()

<StringArray>
['Y']
Length: 1, dtype: str

In [73]:
# Comprobación de valores, "yes" o "no". Tiene nulos
df['OverTime'].unique()

<StringArray>
['Yes', 'No', nan]
Length: 3, dtype: str

In [74]:
# Valor de empresa respecto al empleado. Escala de valores entre 1 y 4, pero solo aparecen dos 
df['PerformanceRating'].unique()

array([3, 4])

In [75]:
# Encuesta de satisfacción. Escala de valores entre 1 y 4. Sin nulos
df['RelationshipSatisfaction'].unique()

array([1, 4, 2, 3])

In [76]:
# Jornada Laboral. Solo presenta dos valores: nulos o 80
df['StandardHours'].unique()

array([80., nan])

In [77]:
# Valores entre 0 y 3. Valor 0 indica sin acciones. Valor 1 nivel alto de acciones. Valor 2 nivel medio de acciones. Valor 3 nivel bajo de acciones
df['StockOptionLevel'].unique()

array([0, 1, 3, 2])

In [78]:
# Valor de formaciones que se ha recibido en el último año. Escala de valores entre 0 y 6. Con nulos
df['TrainingTimesLastYear'].unique()

array([ 0.,  3.,  2.,  5.,  1.,  4., nan,  6.])

In [79]:
# Encuesta de satisfacción. Escala de valores entre 1 y 4. Sin nulos
df['WorkLifeBalance'].unique()

array([1, 3, 2, 4])

**Se identifican las siguientes cuestiones:**

* Las columnas `Attrition`, `Over18` y `OverTime` son valores booleanos en STR. No siguen el mismo formato.
* Las columnas `EmployeeCount`, `Over18`, `PerformanceRating` y `StandardHours` no aportan valor claro al estudio. Es necesario determinar si se descartan valores o se quieren conservar.
* Hay columnas en float que deben pasar a INT: `Age`, `JobSatisfaction`, `StandardHours`, `TrainingTimesLastYear` y `YearsWithCurrManager`.
* Se comprueba que las columnas con encuestas de satisfacción siempre tienen valores del 1 al 4.
* Se encuentra una errata en los valores de `MaritalStatus`.

**Tareas pendientes:**

- [_] Eliminar las columnas que no aportan valor al estudio. Se decide eliminar: `EmployeeCount`, `Over18` y `StandardHours`. `PerformanceRating` se mantiene para posibles estudios posteriores. 
- [_] Cambiar los tipos de columnas: `Age`, `JobSatisfaction`, `StandardHours`, `TrainingTimesLastYear` y `YearsWithCurrManager`a Int64 y `Attrition` y `OverTime` a booleano.
- [_] Ajustar los valores de la columna `MaritalStatus`.
- [_] Usar `EmployeeNumber` como índice del dataset.

## Limpieza y transformación de datos


In [80]:
#Convertimos la columna Age a número entero
## Usamos Int64 en vez de int porque Age tiene valores nulos
# El tipo Int64 permite trabajar con enteros y mantener los nulos como <NA>
df['Age'] = df['Age'].astype('Int64')
df['Age'].dtype

Int64Dtype()

In [81]:
#Convertimos la columna JobSatisfaction a número entero
## Usamos Int64 en vez de int porque JobSatisfaction tiene valores nulos
# El tipo Int64 permite trabajar con enteros y mantener los nulos como <NA>
df['JobSatisfaction'] = df['JobSatisfaction'].astype('Int64')
df['JobSatisfaction'].dtype

Int64Dtype()

In [82]:
#Convertimos la columna StandardHours a número entero, aunque se contempla el descarte
## Usamos Int64 en vez de int porque StandardHours tiene valores nulos
# El tipo Int64 permite trabajar con enteros y mantener los nulos como <NA>
df['StandardHours'] = df['StandardHours'].astype('Int64')
df['StandardHours'].dtype

Int64Dtype()

In [83]:
#Convertimos la columna TrainingTimesLastYear a número entero
## Usamos Int64 en vez de int porque TrainingTimesLastYear tiene valores nulos
# El tipo Int64 permite trabajar con enteros y mantener los nulos como <NA>
df['TrainingTimesLastYear'] = df['TrainingTimesLastYear'].astype('Int64')
df['TrainingTimesLastYear'].dtype

Int64Dtype()

In [84]:
#Convertimos la columna YearsWithCurrManager a número entero
## Usamos Int64 en vez de int porque YearsWithCurrManager tiene valores nulos
# El tipo Int64 permite trabajar con enteros y mantener los nulos como <NA>
df['YearsWithCurrManager'] = df['YearsWithCurrManager'].astype('Int64')
df['YearsWithCurrManager'].dtype

Int64Dtype()

In [85]:
# Convertimos JobRole a título y eliminamos espacios sobrantes al principio y al final de cada valor
df['JobRole'] = df['JobRole'].str.strip().str.title()
df.head(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993.0,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130.0,24907,1,Y,No,23,4,4,<NA>,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090.0,2396,6,Y,Yes,15,3,2,<NA>,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909.0,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468.0,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


In [86]:
# Cambiamos guiones bajos de BusinessTrave por espacios
df['BusinessTravel'] = df['BusinessTravel'].str.replace("_", " ")
df.head(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993.0,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130.0,24907,1,Y,No,23,4,4,<NA>,1,10,3,3,10,7,1,7
2,37,Yes,Travel Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090.0,2396,6,Y,Yes,15,3,2,<NA>,0,7,3,3,0,0,0,0
3,33,No,Travel Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909.0,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468.0,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


In [87]:
# Convertimos la columna Attrition a valores booleanos
# 'Yes' se transforma en True
# 'No' se transforma en False
df['Attrition'] = df['Attrition'].map({
    'Yes': True,
    'No': False
}).astype('boolean')
df['Attrition'].unique()

<BooleanArray>
[True, False]
Length: 2, dtype: boolean

In [88]:
# Comprobación del tipo final
df['Attrition'].dtype

BooleanDtype

In [89]:
# Convertimos la columna OverTime a valores booleanos
# 'Yes' se transforma en True
# 'No' se transforma en False
df['OverTime'] = df['OverTime'].map({'Yes': True, 'No': False}).astype('boolean')

In [90]:
# Comprobación del tipo final
df['OverTime'].dtype

BooleanDtype

In [91]:
# Eliminamos las filas completamente duplicadas
df = df.drop_duplicates(keep='first')
print(df.duplicated().sum())

0


In [92]:
# Eliminación de las columnas que no aportan valor al estudio
df.drop(columns=['EmployeeCount' , 'Over18', 'StandardHours'], inplace=True)

In [93]:
# EmployeeNumber identifica de forma única a cada empleado/a,
# por lo que lo usamos como índice del dataframe.
df = df.set_index('EmployeeNumber')

In [94]:
# Eliminación de la errata localizada
df['MaritalStatus'] = df['MaritalStatus'].replace('Marreid', 'Married')

In [95]:
# Comprobación de los valores actuales
df['MaritalStatus'].value_counts()

MaritalStatus
Married     606
Single      436
Divorced    296
Name: count, dtype: int64

In [96]:
# Visualización final del dataset
df.head(10)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
EmployeeNumber,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
1,41,True,Travel Rarely,1102,Sales,1,2,Life Sciences,2,Female,94,3,2,Sales Executive,4,Single,5993.0,19479,8,True,11,3,1,0,8,0,1,6,4,0,5
2,49,False,Travel Frequently,279,Research & Development,8,1,Life Sciences,3,Male,61,2,2,Research Scientist,2,Married,5130.0,24907,1,False,23,4,4,1,10,3,3,10,7,1,7
4,37,True,Travel Rarely,1373,Research & Development,2,2,Other,4,Male,92,2,1,Laboratory Technician,3,Single,2090.0,2396,6,True,15,3,2,0,7,3,3,0,0,0,0
5,33,False,Travel Frequently,1392,Research & Development,3,4,Life Sciences,4,Female,56,3,1,Research Scientist,3,Married,2909.0,23159,1,True,11,3,3,0,8,3,3,8,7,3,0
7,27,False,Travel Rarely,591,Research & Development,2,1,Medical,1,Male,40,3,1,Laboratory Technician,2,Married,3468.0,16632,9,False,12,3,4,1,6,3,3,2,2,2,2
8,32,False,Travel Frequently,1005,Research & Development,2,2,Life Sciences,4,Male,79,3,1,Laboratory Technician,4,Single,3068.0,11864,0,False,13,3,3,0,8,2,2,7,7,3,6
10,59,False,Travel Rarely,1324,Research & Development,3,3,Medical,3,Female,81,4,1,Laboratory Technician,1,Married,2670.0,9964,4,<NA>,20,4,1,3,12,3,2,1,0,0,<NA>
11,30,False,Travel Rarely,1358,NaN,24,1,Life Sciences,4,Male,67,3,1,Laboratory Technician,3,Divorced,2693.0,13335,1,False,22,4,2,1,1,2,3,1,0,0,0
12,38,False,Travel Frequently,216,Research & Development,23,3,Life Sciences,4,Male,44,2,3,Manufacturing Director,3,Single,9526.0,8787,0,False,21,4,2,0,10,2,3,9,7,1,8


**Tareas finalizadas**:

- [x] Cambiar `JobRol`y `BusinessTravel` para continuidad de la estética
- [x] Eliminación de duplicados.
- [x] Eliminar las columnas que no aportan valor al estudio. Se decide eliminar: `EmployeeCount`, `Over18` y `StandardHours`. `PerformanceRating` se mantiene para posibles estudios posteriores. 
- [x] Cambiar los tipos de columnas: `Age`, `JobSatisfaction`, `StandardHours`, `TrainingTimesLastYear` y `YearsWithCurrManager`a Int64 y `Attrition` y `OverTime` a booleano.
- [x] Ajustar los valores de la columna `MaritalStatus`.
- [x] Usar `EmployeeNumber` como índice del dataset.

**Análisis de los nulos:**

* Se comprueba que no hay columnas con nulos superiores al 25%, aun así se decide automatizar el proceso considerando dichos valores por posibles estudios futuros con un dataset diferente.
* Se verifica que ninguna columna necesita imputación fuera de las reglas habituales.

**Tareas pendientes**:

- [_] Imputación de nulos, según el tipo de columna y los valores.

In [97]:
# Comprobación de tipos correctos en todas las columnas. Gestión de nulos pendientes
df.info()

<class 'pandas.DataFrame'>
Index: 1470 entries, 1 to 2068
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1397 non-null   Int64  
 1   Attrition                 1470 non-null   boolean
 2   BusinessTravel            1353 non-null   str    
 3   DailyRate                 1470 non-null   int64  
 4   Department                1441 non-null   str    
 5   DistanceFromHome          1470 non-null   int64  
 6   Education                 1470 non-null   int64  
 7   EducationField            1412 non-null   str    
 8   EnvironmentSatisfaction   1470 non-null   int64  
 9   Gender                    1470 non-null   str    
 10  HourlyRate                1470 non-null   int64  
 11  JobInvolvement            1470 non-null   int64  
 12  JobLevel                  1470 non-null   int64  
 13  JobRole                   1470 non-null   str    
 14  JobSatisfaction         

In [98]:
# Función para gestión de nulos de las columnas categóricas
def sustitucion_nulos_categoricas(df):
    # Copia del DF para evitar suscribir los datos en el original
    df = df.copy()
    # Selección de las columnas categóricas
    columnas_categoricas = df.select_dtypes(include='object').columns
    # Filtrado de las columnas con nulos
    columnas_con_nulos = [col for col in columnas_categoricas if df[col].isnull().sum() > 0]
    for columna in columnas_con_nulos:
        porcentaje_nulos = df[columna].isnull().sum()/df.shape[0]*100
        print(f"{columna}. Nulos antes = {df[columna].isnull().sum()} ({porcentaje_nulos:.1f}%)")

        # Búsqueda del valor dominante en la columna
        lista_valores = df[columna].value_counts()/df.shape[0]*100
        dominante = lista_valores.iloc[0]

        # Acción para valores por encima de 25% se imputa con la moda
        if porcentaje_nulos > 25:
            df[columna] = df[columna].fillna(lista_valores.index[0])
            print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")

        # Por debajo de 25% se comprueba la dominancia
        else:
            print(f"{columna}. Dominante = {dominante:.1f}%")
            
            # Si la categoría dominante es inferior al 75% creamos una nueva categoría
            if dominante <= 75:
                df[columna] = df[columna].fillna('unknown')
                print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")

            # Si la categoría dominante es superior al 75% se imputa con la moda    
            else:
                df[columna] = df[columna].fillna(lista_valores.index[0])
                print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")

    return df

In [99]:
# Ejecución de la función e imputación de las columnas categóricas
df_cat_limpio = sustitucion_nulos_categoricas(df)
df_cat_limpio.sample(5)

BusinessTravel. Nulos antes = 117 (8.0%)
BusinessTravel. Dominante = 64.7%
BusinessTravel. Nulos después = 0
Department. Nulos antes = 29 (2.0%)
Department. Dominante = 63.9%
Department. Nulos después = 0
EducationField. Nulos antes = 58 (3.9%)
EducationField. Dominante = 39.5%
EducationField. Nulos después = 0
MaritalStatus. Nulos antes = 132 (9.0%)
MaritalStatus. Dominante = 41.2%
MaritalStatus. Nulos después = 0


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
EmployeeNumber,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
1648,27,False,Travel Frequently,591,Research & Development,2,3,Medical,4,Male,87,3,1,Research Scientist,4,Single,2580.0,6297,2,False,13,3,3,0,6,0,2,4,2,1,2
869,28,False,Travel Rarely,1169,Human Resources,8,2,Medical,2,Male,63,2,1,Human Resources,4,Divorced,4936.0,23965,1,False,13,3,4,1,6,6,3,5,1,0,4
1592,23,False,Travel Rarely,977,Research & Development,10,3,Technical Degree,4,Male,45,4,1,Research Scientist,3,Married,2073.0,12826,2,False,16,3,4,1,4,<NA>,3,2,2,2,2
714,27,False,Travel Frequently,1410,Sales,3,1,Medical,4,Female,71,4,2,Sales Executive,4,Divorced,4647.0,16673,1,True,20,4,2,2,6,3,3,6,5,0,4
657,20,False,Travel Rarely,959,Research & Development,1,3,Life Sciences,4,Female,83,2,1,Research Scientist,2,Single,2836.0,11757,1,False,13,3,4,0,1,0,4,1,0,0,0


In [100]:
# Función para gestión de nulos de las columnas booleanas
def sustitucion_nulos_booleanas(df):
    # Copia del DF para evitar suscribir los datos en el original
    df = df.copy()
    # Selección de las columnas booleanas
    columnas_bool = df.select_dtypes(include='boolean').columns
    # Filtrado de columnas con nulos
    columnas_con_nulos = [col for col in columnas_bool if df[col].isnull().sum() > 0]

    for columna in columnas_con_nulos:
        moda = df[columna].mode()[0]
        print(f"{columna}: nulos antes = {df[columna].isnull().sum()}. Se sustituye con la moda: ({moda})")
        # Imputación con la moda
        df[columna] = df[columna].fillna(moda)

    return df

In [101]:
# Ejecución de la función e imputación de las columnas booleanas
df_bool_limpio = sustitucion_nulos_booleanas(df_cat_limpio)
df_bool_limpio.sample(5)

OverTime: nulos antes = 44. Se sustituye con la moda: (False)


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
EmployeeNumber,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
1423,58,False,Travel Rarely,1055,Research & Development,1,3,Medical,4,Female,76,3,5,Research Director,1,Married,19701.0,22456,3,True,21,4,3,1,32,3,3,9,8,1,5
303,54,False,Travel Rarely,1147,Sales,3,3,Marketing,4,Female,52,3,2,Sales Executive,1,Married,5940.0,17011,2,False,14,3,4,1,16,4,3,6,2,0,5
667,27,True,Travel Rarely,1420,Sales,2,1,Marketing,3,Male,85,3,1,Sales Representative,1,Divorced,3041.0,16346,0,False,11,3,2,1,5,3,3,4,3,0,2
98,31,False,Travel Rarely,746,Research & Development,8,4,Life Sciences,3,Female,61,3,2,Manufacturing Director,4,Single,4424.0,20682,1,False,23,4,4,0,11,2,3,11,7,1,8
1268,33,False,unknown,536,Sales,10,5,Marketing,4,Male,82,4,3,Sales Executive,3,Divorced,8380.0,21708,0,True,14,3,4,2,10,3,3,9,8,0,8


In [102]:
# Función para gestión de nulos de las columnas numéricas
def sustitucion_nulos_numericas(df):
    # Copia del DF para evitar suscribir los datos en el original
    df = df.copy()
    # Selección de las columnas numéricas
    columnas_numericas = df.select_dtypes(include=['number']).columns
    # Filtrado de columnas con nulos
    columnas_con_nulos = [col for col in columnas_numericas if df[col].isnull().sum() > 0]
    # Entrenamiento del imputador
    imputador = IterativeImputer(max_iter= 10, random_state= 42)
    valores_imputados = imputador.fit_transform(df[columnas_numericas])
    df_temp = pd.DataFrame(valores_imputados, columns=columnas_numericas, index=df.index)
    
    for columna in columnas_con_nulos:
        porcentaje_nulos = df[columna].isnull().sum()/df.shape[0]*100
        print(f"{columna}. Nulos antes = {df[columna].isnull().sum()} ({porcentaje_nulos:.1f}%)")

        # En nulos superiores a un 25% se sustituye con valores del imputador
        if porcentaje_nulos > 25:
            df[columna] = df_temp[columna]
            print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")

        # En nulos inferiores a un 25% se considera si hay sesgo entre media y mediana
        else:
            sesgo = df[columna].skew()
            mediana = df[columna].median()
            media = df[columna].mean() 
            print(f"{columna}. Sesgo = {sesgo:.2f}")

            # Para sesgos superiores a un 0.5 se imputa con la mediana
            if abs(sesgo) > 0.5:
                df[columna] = df[columna].fillna(round(mediana))
                print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")
                
            # Para sesgos inferiores a un 0.5 se imputa con la media
            else:
                df[columna] = df[columna].fillna(round(media))
                print(f"{columna}. Nulos después = {df[columna].isnull().sum()}")

    return df

In [103]:
# Ejecución de la función e imputación de las columnas numéricas
df_limpio = sustitucion_nulos_numericas(df_bool_limpio)
df_limpio.sample(5)

Age. Nulos antes = 73 (5.0%)
Age. Sesgo = 0.41
Age. Nulos después = 0
JobSatisfaction. Nulos antes = 29 (2.0%)
JobSatisfaction. Sesgo = -0.33
JobSatisfaction. Nulos después = 0
MonthlyIncome. Nulos antes = 14 (1.0%)
MonthlyIncome. Sesgo = 1.37
MonthlyIncome. Nulos después = 0
TrainingTimesLastYear. Nulos antes = 88 (6.0%)
TrainingTimesLastYear. Sesgo = 0.55
TrainingTimesLastYear. Nulos después = 0
YearsWithCurrManager. Nulos antes = 147 (10.0%)
YearsWithCurrManager. Sesgo = 0.85
YearsWithCurrManager. Nulos después = 0


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
EmployeeNumber,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
401,26,True,Travel Rarely,950,Sales,4,4,Marketing,4,Male,48,2,2,Sales Executive,4,Single,5828.0,8450,1,True,12,3,2,0,8,0,3,8,7,7,4
497,27,False,unknown,1469,Research & Development,1,2,Medical,4,Male,82,3,1,Laboratory Technician,2,Divorced,3816.0,17881,1,False,11,3,2,1,5,2,3,5,2,0,4
806,30,False,Travel Rarely,921,Research & Development,1,3,Life Sciences,4,Male,38,1,1,Laboratory Technician,3,Married,3833.0,24375,3,False,21,4,3,2,7,2,3,2,2,0,2
478,21,True,Travel Frequently,756,Sales,1,1,Technical Degree,1,Female,99,2,1,Sales Representative,2,Single,2174.0,9150,1,True,11,3,3,0,3,3,3,3,2,1,2
281,40,False,unknown,905,Research & Development,19,2,Medical,3,Male,99,3,2,Laboratory Technician,4,Married,2741.0,16523,8,True,15,3,3,1,15,2,4,7,2,3,7


In [104]:
# Verificación de que no existen nulos en el dataset
df_limpio.isnull().any().any()

np.False_

In [105]:
# Última comprobación de la información del dataset
df_limpio.info()

<class 'pandas.DataFrame'>
Index: 1470 entries, 1 to 2068
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1470 non-null   Int64  
 1   Attrition                 1470 non-null   boolean
 2   BusinessTravel            1470 non-null   str    
 3   DailyRate                 1470 non-null   int64  
 4   Department                1470 non-null   str    
 5   DistanceFromHome          1470 non-null   int64  
 6   Education                 1470 non-null   int64  
 7   EducationField            1470 non-null   str    
 8   EnvironmentSatisfaction   1470 non-null   int64  
 9   Gender                    1470 non-null   str    
 10  HourlyRate                1470 non-null   int64  
 11  JobInvolvement            1470 non-null   int64  
 12  JobLevel                  1470 non-null   int64  
 13  JobRole                   1470 non-null   str    
 14  JobSatisfaction         

## Creación de csv final limpio:

Tras la transformación y limpieza, el dataset está listo para la fase de análisis estadístico y visualización.
Datos finales:

* Filas: 1474
* Columnas: 35
* Nulos: 0
* Duplicados: 0

Con estos datos, se crea un nuevo CSV con los valores transformados

In [106]:
# Creación del csv limpio
df_limpio.to_csv('files/hr_final.csv')